lib/update/src/lib.rs

```diff
@@ -0,0 +1,51 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+// Device independent helpers to do image update
+#![no_std]
+
+pub use drv_update_api::UpdateError;
+use userlib::TaskId;
+
+// Size of processing for the driver. A driver may pass less than this
+// size but it cannot pass greater than this size.
+pub fn block_size(update_task: TaskId) -> Result<usize, UpdateError> {
+    let update = drv_update_api::Update::from(update_task);
+
+    update.block_size()
+}
+
+// Start of an image update. Must be called before trying to write a block
+// TODO pass the header?
+pub fn start_update(update_task: TaskId) -> Result<(), UpdateError> {
+    let update = drv_update_api::Update::from(update_task);
+
+    update.prep_image_update()
+}
+
+// Write data to the specified block number. It is expected that the driver
+// has called the start function before this.
+//
+// The data must be at least `block_size` in length.
+pub fn write_block(
+    update_task: TaskId,
+    start_block: usize,
+    data: &[u8],
+) -> Result<(), UpdateError> {
```

Comment:
I have a few questions about this API, but for context, do updates write the contiguous blocks `0..sizeof(update)/block_size`, or is it more scattered? I'm assuming it's the former; if that's wrong these questions might make less sense.

1. Does passing a `start_block` of anything other than 0 require the caller to also call `block_size()` first so they can convert whatever offset they have (maybe the number of bytes written so far?) into a block number?
2. Does `data.len()` need to be a multiple of the block size? (If not, why does it need to be at least `block_size` in length?)
3. Should this function return the number of blocks written? Returning a `usize` might be confusing at a glance (is it "number of blocks" or "number of bytes"), but I think that potential confusion is related to all of these questions.

Reply:
Correct, the expectation right now is that we'll be writing the blocks contiguously. We don't really have a use for doing scattered.

1. Correct
2. It does not need to be a multiple. It's actually the opposite for the `write_one_block` call, the data can be at most a block size (was there a check elsewhere that had the constraint?). This is a limitation of the current design with leases where we need to copy the data from the other task into a fixed size buffer before we can write it out. There might be some room for improvement.
3. What use would you see in returning the number of block written?

Reply:
On 3 - Assuming I'm calling this function multiple times to write ultimately-contiguous data, I have to do some bookkeeping to know what `start_block` should be for 2nd-and-later calls; this function could return it directly since it already knows. That said, maybe this doesn't actually help, because if I'm calling this multiple times I already have to know that `data` is an exact multiple of `block_size` (unless it's the final block), right? I'm trying to picture what streaming an update in over UDP looks like; assuming those packets aren't delivering data that's an exact multiple of the block size, the task receiving them is already going to have to know the block size to do some buffering, etc.

I think the thing that's bugging me about this is that it's taking a start block explicitly and a number of blocks implicitly (`data.len() / block_size`, where `block_size` is determined internally by this function). The caller can independently ask what the block size this function is going to use is, but that seems slightly awkward.

Reply:
The example in the README was what I had in mind how this would work
```
prep_image_update()
for (i, block) in image_byte_array {
	write_one_block(i, block)
}
finish_image_update()
```
Your example of receiving an update over UDP (which is much closer to how this will actually work) shows the weaknesses of this design. I do think the `write_one_block` which takes a block number and data is the right approach at the lowest level, the question is then what to build on top of it.

I think having a structure that you can feed data without having to worry about the block number would be easier to use. That unfortunately doesn't work with doing things through hiffy because there isn't an actual task where you can keep the structure. Let me give this some thought and see if I can come up with a design/commenting that's easier to use.

Reply:
It also might be perfectly reasonable to not provide a block-number-free abstraction at this point, and let the task receiving an update over UDP worry about it, especially if that's the only place we think it will be used. At that point I might argue for dropping this function altogether and only providing `write_one_block`?

Reply:
based on other suggestions from mkeeter I dropped this library all together and just folded the implementation in hiffy for now. Once we have a few more details about how updates are going to come in we can evaluate what's a good abstraction for state.